src/aapns/pool.py

```diff
@@ -0,0 +1,219 @@
+        bits.append(f"retrying:{self.retrying}")
+        bits.append(f"completed:{self.completed}")
+        bits.append(f"errors:{self.errors}")
+        return "<Pool %s>" % " ".join(bits)
+
+    def __init__(self, origin: str, size=2, ssl=None):
+        self.origin = origin
+        self.conn: Set[Connection] = set()
+        self.dying: Set[Connection] = set()
+        self.ssl = ssl if ssl else create_ssl_context()
+        assert size > 0
+        self.size = size
+        self._size_event = Event()
+
+    @property
+    def state(self):
+        if not self.bg:
+            return "new"
+        elif not self.started:
+            return "starting"
+        elif not self.closing:
+            return "active"
+        elif not self.closed:
+            return "closing"
+        else:
+            return "closed"
+
+    def resize(self, size):
+        assert size > 0
+        self.size = size
+        self._size_evet.set()
+
+    def cert_hook(self, c):
+        if not self.outcome and c.outcome == "BadCertificateEnvironment":
+            self.closing = True
+            self.outcome = c.outcome
+
+    async def background_resize(self):
+        while True:
+            if self.closing or self.closed:
+                return
+
+            for c in list(self.conn):
+                if c.closing:
+                    self.conn.remove(c)
+                    self.dying.add(c)
+                    self.cert_hook(c)
+
+            while len(self.conn) > self.size:
+                c = self.conn.pop()
+                c.closing = True
+                self.dying.add(c)
+                self.cert_hook(c)
+
+            for c in list(self.dying):
+                if c.closed:
+                    self.dying.remove(c)
+                    self.cert_hook(c)
+                elif not c.channels:
+                    self.dying.remove(c)
+                    try:
+                        await c.__aexit__(None, None, None)
+                    finally:
+                        self.cert_hook(c)
+                if self.closing or self.closed:
+                    return
+
+            while len(self.conn) < self.size:
+                if not await self.add_one_connection():
+                    break
+                if self.closing or self.closed:
+                    return
+
+            # FIXME wait for a trigger:
+            # * some connection state has changed
+            with suppress(TimeoutError):
+                await wait_for(self._size_event.wait(), timeout=1)
+            self._size_event.clear()
+
+    async def add_one_connection(self):
+        try:
+            c = Connection(self.origin, ssl=self.ssl)
+            await c.__aenter__()
+            self.conn.add(c)
+            self.cert_hook(c)
+            return True
+        except Exception:
+            logger.exception("Failed creating APN connection")
+            self.cert_hook(c)
+
+    async def __aenter__(self):
```

Comment:
same as for `Connection`, to make it harder to re-use/arguably easier to understand, make an async classmethod to initialize a pool and have `__init__` receive all the bits it needs to work already.